scripts-dev/complement.sh

```diff
@@ -45,5 +45,5 @@ if [[ -n "$1" ]]; then
   EXTRA_COMPLEMENT_ARGS+="-run $1 "
 fi
 
-# Run the tests!
-COMPLEMENT_BASE_IMAGE=complement-synapse go test -v -tags synapse_blacklist,msc2946,msc3083 -count=1 $EXTRA_COMPLEMENT_ARGS ./tests
+# Run the tests on the resulting image!
+COMPLEMENT_BASE_IMAGE=complement-synapse go test -tags msc2716 -v -count=1 ./tests/main_test.go ./tests/msc2716_test.go -run TestBackfillingHistory/parallel/Backfilled_historical_events_resolve_with_proper_state
```

Comment:
I added a Complement test for whether the historical messages can federate, `Historical messages are visible on federated server`.

But the test is currently failing because the `/backfill` request is only returning 7 events instead of all 8 (with the historical message). Homeserver logs here: https://github.com/matrix-org/complement/pull/68/files#r620798987

The test can be summarized as:

 1. Room is created
 1. Alice joins the room
 1. Alice sends a message in the room
 1. The application service backfills a historical message from Maria
 1. Then remote Charlie from `hs2` joins the room
 1. `hs1` and `hs2` talk to each other to backfill (only 7 events passed across)
 1. Charlie does a `/messages` and only gets 7 events back and is missing the historical message

Reply:
My first suspicion was because the `/backfill` request, query param `v` was the `eventBefore` which was the latest "real" message and our historical message actually came after, we weren't seeing it. Not sure how it chooses `v` though.

But then I added some `eventsAfter` so it would have some buffer in between and hopefully navigate up the DAG to the historical message but it still never reaches it.


 1. Room is created
 1. Alice joins the room
 1. Alice sends a message in the room (`eventBefore`)
 1. More messages happen in the room (`eventsAfter`)
 1. The application service backfills a historical message from Maria (next to the message from Alice)
 1. Then remote Charlie from `hs2` joins the room
 1. `hs1` and `hs2` talk to each other to backfill (only 7 events passed across)
 1. Charlie does a `/messages` and only gets 7 events back and is missing the historical message

---

I'm guessing the historical event isn't being put in the `event_edges` which the federation query is looking at, [`synapse/storage/databases/main/event_federation.py#L883-L890` -> `_get_backfill_events`](https://github.com/matrix-org/synapse/blob/3ff225175462dde8376aa584e3a47c43b1f0e790/synapse/storage/databases/main/event_federation.py#L883-L890). A historical message with the same depth isn't an edge of the subsequent "real" message?

 - `eventBefore` "real" message: `$s5UizgbmQ9FxKUlls2j3FpHfGDcQIepmHwukiLAe7xI`
 - `eventsAfter` "real" messages: `$gVLfz7jUk1FmJw60ZDREusUPrFfEbncyuG6p68hVi10`, `$XCxnST7JmRI-Z0z4rA8FNDMjupXbpX9VplWuiI5tG5s`, `$ADdt9L6aZLryad0jHuYutKfMSby7G76wVkDMTaNumPI`
 - Historical message: `$ygOwEi6CHxmG2_JU74h81jVl8hi6xoJWpNTtujZlEKc`
```
2021-04-27 06:53:55,390 - synapse.federation.transport.server - 470 - INFO - GET-18 - FederationBackfillServlet.on_GET query={b'v': [b'$gVLfz7jUk1FmJw60ZDREusUPrFfEbncyuG6p68hVi10'], b'limit': [b'100']}
2021-04-27 06:53:55,390 - synapse.federation.transport.server - 480 - INFO - GET-18 - FederationBackfillServlet.on_GET versions=1 -> ['$gVLfz7jUk1FmJw60ZDREusUPrFfEbncyuG6p68hVi10']
2021-04-27 06:53:55,390 - synapse.federation.federation_server - 150 - INFO - GET-18 - federation_server.on_backfill_request versions=1 -> ['$gVLfz7jUk1FmJw60ZDREusUPrFfEbncyuG6p68hVi10']
2021-04-27 06:53:55,392 - synapse.handlers.federation - 1953 - INFO - GET-18 - handlers.on_backfill_request pdu_list 1 ['$gVLfz7jUk1FmJw60ZDREusUPrFfEbncyuG6p68hVi10']
2021-04-27 06:53:55,392 - synapse.storage.databases.main.event_federation - 918 - INFO - GET-18 - _get_backfill_events: !YQMLneJZggFmBKENOs:hs1, ['$gVLfz7jUk1FmJw60ZDREusUPrFfEbncyuG6p68hVi10'], 100
2021-04-27 06:53:55,393 - synapse.storage.databases.main.event_federation - 959 - INFO - GET-18 - _get_backfill_events query: $gVLfz7jUk1FmJw60ZDREusUPrFfEbncyuG6p68hVi10
2021-04-27 06:53:55,393 - synapse.storage.databases.main.event_federation - 963 - INFO - GET-18 - _get_backfill_events query row: (9, '$XCxnST7JmRI-Z0z4rA8FNDMjupXbpX9VplWuiI5tG5s')
2021-04-27 06:53:55,393 - synapse.storage.databases.main.event_federation - 959 - INFO - GET-18 - _get_backfill_events query: $XCxnST7JmRI-Z0z4rA8FNDMjupXbpX9VplWuiI5tG5s
2021-04-27 06:53:55,393 - synapse.storage.databases.main.event_federation - 963 - INFO - GET-18 - _get_backfill_events query row: (8, '$ADdt9L6aZLryad0jHuYutKfMSby7G76wVkDMTaNumPI')
2021-04-27 06:53:55,393 - synapse.storage.databases.main.event_federation - 959 - INFO - GET-18 - _get_backfill_events query: $ADdt9L6aZLryad0jHuYutKfMSby7G76wVkDMTaNumPI
2021-04-27 06:53:55,393 - synapse.storage.databases.main.event_federation - 963 - INFO - GET-18 - _get_backfill_events query row: (7, '$s5UizgbmQ9FxKUlls2j3FpHfGDcQIepmHwukiLAe7xI')
2021-04-27 06:53:55,393 - synapse.storage.databases.main.event_federation - 959 - INFO - GET-18 - _get_backfill_events query: $s5UizgbmQ9FxKUlls2j3FpHfGDcQIepmHwukiLAe7xI
2021-04-27 06:53:55,394 - synapse.storage.databases.main.event_federation - 963 - INFO - GET-18 - _get_backfill_events query row: (6, '$X-Qvb69TT7d-1BBWBkUvg1BJ1zupZQginmtPCZUG60s')
2021-04-27 06:53:55,394 - synapse.storage.databases.main.event_federation - 959 - INFO - GET-18 - _get_backfill_events query: $X-Qvb69TT7d-1BBWBkUvg1BJ1zupZQginmtPCZUG60s
2021-04-27 06:53:55,394 - synapse.storage.databases.main.event_federation - 963 - INFO - GET-18 - _get_backfill_events query row: (5, '$_4YADPiAh44toXm84EFSIFiEVrPyQfffDM-nEB-1fb4')
2021-04-27 06:53:55,394 - synapse.storage.databases.main.event_federation - 959 - INFO - GET-18 - _get_backfill_events query: $_4YADPiAh44toXm84EFSIFiEVrPyQfffDM-nEB-1fb4
2021-04-27 06:53:55,394 - synapse.storage.databases.main.event_federation - 963 - INFO - GET-18 - _get_backfill_events query row: (4, '$jmc8zyHEYP47LZ8I-0UtoojRDN74445XcGn4EGn5RZE')
2021-04-27 06:53:55,394 - synapse.storage.databases.main.event_federation - 959 - INFO - GET-18 - _get_backfill_events query: $jmc8zyHEYP47LZ8I-0UtoojRDN74445XcGn4EGn5RZE
2021-04-27 06:53:55,394 - synapse.storage.databases.main.event_federation - 963 - INFO - GET-18 - _get_backfill_events query row: (3, '$kOwEb1YLkMFoLfmfvPlPBuo_jX2n5UhX4bnDKdmNvGA')
2021-04-27 06:53:55,394 - synapse.storage.databases.main.event_federation - 959 - INFO - GET-18 - _get_backfill_events query: $kOwEb1YLkMFoLfmfvPlPBuo_jX2n5UhX4bnDKdmNvGA
2021-04-27 06:53:55,395 - synapse.storage.databases.main.event_federation - 963 - INFO - GET-18 - _get_backfill_events query row: (2, '$pSVBAkvtFjOhKCx1FlTa4GmsVA35l_ft7fH63v4i720')
2021-04-27 06:53:55,395 - synapse.storage.databases.main.event_federation - 959 - INFO - GET-18 - _get_backfill_events query: $pSVBAkvtFjOhKCx1FlTa4GmsVA35l_ft7fH63v4i720
2021-04-27 06:53:55,395 - synapse.storage.databases.main.event_federation - 963 - INFO - GET-18 - _get_backfill_events query row: (1, '$FTtSmuF5XY_th35z2FzJzmJzrlGdIkbWvAjYphn_WpE')
2021-04-27 06:53:55,395 - synapse.storage.databases.main.event_federation - 959 - INFO - GET-18 - _get_backfill_events query: $FTtSmuF5XY_th35z2FzJzmJzrlGdIkbWvAjYphn_WpE
2021-04-27 06:53:55,395 - synapse.handlers.federation - 1959 - INFO - GET-18 - handlers.on_backfill_request 10
2021-04-27 06:53:55,396 - synapse.handlers.federation - 1963 - INFO - GET-18 - handlers.on_backfill_request after filter 10
2021-04-27 06:53:55,398 - synapse.access.https.8448 - 372 - INFO - GET-18 - ::ffff:172.27.0.2 - 8448 - {hs2} Processed request: 0.009sec/-0.000sec (0.007sec, 0.000sec) (0.001sec/0.003sec/2) 6665B 200 "GET /_matrix/federation/v1/backfill/%21YQMLneJZggFmBKENOs%3Ahs1?v=%24gVLfz7jUk1FmJw60ZDREusUPrFfEbncyuG6p68hVi10&limit=100 HTTP/1.1" "Synapse/1.31.0" [0 dbevts]
```

![](https://user-images.githubusercontent.com/558581/116199404-b3983c00-a6fc-11eb-8573-12d0f8177ff1.png)

Reply:
@erikjohnston Can I get some help on understanding what an "edge" is? And if you spot it, why the historical message is not part of the `/backfill` response?

Reply:
(re: "Edge", in graph theory, an 'edge' is the line between two nodes)

Reply:
Thanks @ShadowJonathan 😺

Reply:
What are the prev events of the join of `hs2`? Does anything reference the historical event? If not then `/backfill`, which currently just walks up the ancestors of the supplied events, won't return it. We'll probably want to extend the `_get_backfill_events` to also pull in these "new" historic events as well.

The other issue that's likely to crop up here is what happens when a server joins, gets *all* the history, and *then* the original HS adds some historic events. The remote server won't ever call `/backfill` again as it thinks it has all the events in the room (since it won't have anything in its `event_backward_extremities` table, which is what governs when and with what to call `/backfill` with).

I think what we'd want to do is to force a gap at where we'd want to insert history, that way whenever another server reaches that point it'll always try and backfill there. When the `_get_backfill` function is called on the original server it knows that if it reaches that gap to go and pull in another chunk of historical events, iyswim? Downside to this is that only the original server would know to how to handle `/backfill` there.

We could also add an event there named "insertion point" which contains the server and an opaque ID for the next "chunk" of historical events to be pulled. I.e. when a server reaches that point, and hasn't yet received the chunk, it can go and request it from the server (which could then dynamically load it in). Or something

Reply:
> What are the prev events of the join of `hs2`? Does anything reference the historical event? If not then `/backfill`, which currently just walks up the ancestors of the supplied events, won't return it.

Ahh, this explains how the `v` parameter is set :+1:

---

> We'll probably want to extend the `_get_backfill_events` to also pull in these "new" historic events as well.

Thanks for the tip! I'll look into making it traverse the successors looking for `m.historical`

---

> The other issue that's likely to crop up here is what happens when a server joins, gets all the history, and then the original HS adds some historic events. 

Can we use the push part of the federation whenever new historical messages come in? (hs1 calls `PUT hs2/_matrix/federation/v1/send/{txnId}`)

Seems like if `hs2` is already joined to the room, we can just fill them in when `hs1` gets historical messages added as well. And then this avoid any need for "gaps" or "insertion points"

---

> I think what we'd want to do is to force a gap at where we'd want to insert history, that way whenever another server reaches that point it'll always try and backfill there

@erikjohnston  How do we create a "gap"?

My mind is going to erasing or changing the `prev_event` of the existing "live" message after our historical chunk but that doesn't seem possible.

Reply:
> > The other issue that's likely to crop up here is what happens when a server joins, gets all the history, and then the original HS adds some historic events.
> 
> Can we use the push part of the federation whenever new historical messages come in? (hs1 calls `PUT hs2/_matrix/federation/v1/send/{txnId}`)
> 
> Seems like if `hs2` is already joined to the room, we can just fill them in when `hs1` gets historical messages added as well. And then this avoids any need for "gaps" or "insertion points"

We tend towards *not* having all history on all servers, so I'm not hugely keen to actively push historical stuff that half the servers probably don't care about. It's an option though.

---

> > I think what we'd want to do is to force a gap at where we'd want to insert history, that way whenever another server reaches that point it'll always try and backfill there
> 
> @erikjohnston How do we create a "gap"?
> 
> My mind is going to erasing or changing the `prev_event` of the existing "live" message after our historical chunk but that doesn't seem possible.

The easiest way of creating a gap is by specifying a prev event that doesn't exist, such as `$0`. It's hacky though..

Reply:
> We tend towards not having all history on all servers, so I'm not hugely keen to actively push historical stuff that half the servers probably don't care about. It's an option though.

Using push federation feels like the least complex solution.

The gap and insertion point also feel weird to me since they rely on the original homeserver as the central source of truth which goes against the whole decentralized ethos.


---


I got the historical messages as part of the `/backfill` response but they are failing to auth.

After `hs2` calls `matrix://hs1/backfill`, it's missing the outlier member state event to auth the historical messages, so it calls `matrix://hs1/_matrix/federation/v1/event_auth/<room_id>/<historic_event_id>` to fetch the missing auth event (as expected). The response includes a `m.room.member` event for the historical message author, `@maria:hs1`, but it's a totally different ID than expected and I have no idea where this event is coming from!

If I check the `@maria:hs1` member state event when we first `/bulksend`, it's `$yg2qSK4t4I7_PaPAjbf5kqmIGZobHnu4yjzK2BpBq2M`. But when I compare that to the `/event_auth` response, it has the `@maria:hs1` join as `$HJ1S9pSyk0IayKDIUx1o8YseQDhR5OqwDSLaWAhF0T8` which is totally different and unexpected.

`/bulksend` `auth_events` for the historical message, 7 state events. These auth events are even the same if I run a query after we bulk insert everything  to lookup the historical message by `event_id` .
```diff
[
  '$B2XdB4JXj97qtVbvHN0OJr7kKdtDVDwYGCPpHdU-GsI',
  '$sd7XJzwB6lFcy_pC512072QWlexAy1P4H0MDWFjQD5s',
  '$fV42Jv1SCvuMwv9Sj0pmdZhV9u0WaxEs1tMqs-FGSrY',
  '$-YRnU5-ieM5027RC21oG1US9QwAsmLMzq0SdJzvyDTg',
  '$Q55_Jem_AB_zkZIHVUpnPMU6Acl1n57YQMB4Y_o1rJQ',
  '$Z_kM2eubQ4zJpx4xT2hvo7s7-vDzl1blyrS8z05v5Js',
  # Expected @maria:hs1 join event
+  '$yg2qSK4t4I7_PaPAjbf5kqmIGZobHnu4yjzK2BpBq2M'
]
```


`/event_auth` `auth_event_chain` for the historical message, 7 state events but after validation goes to 6 events. The root of the auth problem seems to be this rogue `@maria:hs1` join.
```diff
[
  <FrozenEventV3 event_id='$Z_kM2eubQ4zJpx4xT2hvo7s7-vDzl1blyrS8z05v5Js', type='m.room.member', state_key='@alice:hs1'>,
  <FrozenEventV3 event_id='$-YRnU5-ieM5027RC21oG1US9QwAsmLMzq0SdJzvyDTg', type='m.room.history_visibility', state_key=''>,
  <FrozenEventV3 event_id='$fV42Jv1SCvuMwv9Sj0pmdZhV9u0WaxEs1tMqs-FGSrY', type='m.room.join_rules', state_key=''>,
  <FrozenEventV3 event_id='$sd7XJzwB6lFcy_pC512072QWlexAy1P4H0MDWFjQD5s', type='m.room.power_levels', state_key=''>,
  <FrozenEventV3 event_id='$Q55_Jem_AB_zkZIHVUpnPMU6Acl1n57YQMB4Y_o1rJQ', type='m.room.member', state_key='@the-bridge-user:hs1'>,
  # This is the random @maria:hs1 join event that does not match what we set in `/bulksend`
  # This event also fails the signature check, "Signature was forged or corrupt"
-  <FrozenEventV3 event_id='$HJ1S9pSyk0IayKDIUx1o8YseQDhR5OqwDSLaWAhF0T8', type='m.room.member', state_key='@maria:hs1'>,
  <FrozenEventV3 event_id='$B2XdB4JXj97qtVbvHN0OJr7kKdtDVDwYGCPpHdU-GsI', type='m.room.create', state_key=''>
]
```



Next step is to investigate where the random `@maria:hs1` join event is coming from and why the one I expect isn't coming back on `/event_auth`.


---



<details>

<summary>Relevant homeserver logs</summary>

```
eventIDsFromResponse="[$3zmsrY2NV1lZuTST2DjsAYnf_n61mj-F7FzwvOaurbM (Message 2) $MzJyjtpWn1VUYgwU_KDKxaXP-4A-SFzl5aPdCpoMtIY (Message 1) $_8vQdgFMafAaWskZoYTdvKFMp44hvS50Myp1DoMyMRg (Message 0) $QWV5wDAv_0JtTk81twPMMOCmjoN-T1egDbJs8WmuAOg (Message 0)]" 

historicalEvents="[$uXlTwCF3DV2WSy5iCG9SKyR-GZFPA6m166ENJ7wtyLY $e_fZOPvlWVTfCnckdZT8zuiT_zArA7rHgf5QKconAc8]" 

historicalStateEvents="[$B2XdB4JXj97qtVbvHN0OJr7kKdtDVDwYGCPpHdU-GsI $sd7XJzwB6lFcy_pC512072QWlexAy1P4H0MDWFjQD5s $fV42Jv1SCvuMwv9Sj0pmdZhV9u0WaxEs1tMqs-FGSrY $-YRnU5-ieM5027RC21oG1US9QwAsmLMzq0SdJzvyDTg $Q55_Jem_AB_zkZIHVUpnPMU6Acl1n57YQMB4Y_o1rJQ $Z_kM2eubQ4zJpx4xT2hvo7s7-vDzl1blyrS8z05v5Js $yg2qSK4t4I7_PaPAjbf5kqmIGZobHnu4yjzK2BpBq2M]" 

localEventIDsFromResponse="[$3zmsrY2NV1lZuTST2DjsAYnf_n61mj-F7FzwvOaurbM (Message 2) $MzJyjtpWn1VUYgwU_KDKxaXP-4A-SFzl5aPdCpoMtIY (Message 1) $_8vQdgFMafAaWskZoYTdvKFMp44hvS50Myp1DoMyMRg (Message 0) $uXlTwCF3DV2WSy5iCG9SKyR-GZFPA6m166ENJ7wtyLY (Historical 1 (chunk=0)) $e_fZOPvlWVTfCnckdZT8zuiT_zArA7rHgf5QKconAc8 (Historical 0 (chunk=0)) $QWV5wDAv_0JtTk81twPMMOCmjoN-T1egDbJs8WmuAOg (Message 0)]"
```



## `hs1`

```
2021-05-01 04:14:15,327 - synapse.rest.client.v1.room - 345 - INFO - POST-12 - /bulksend member state event_id=$yg2qSK4t4I7_PaPAjbf5kqmIGZobHnu4yjzK2BpBq2M state_key=@maria:hs1, content={'membership': 'join'}
2021-05-01 04:14:15,364 - synapse.rest.client.v1.room - 402 - INFO - POST-12 - /bulksend event_id=$uXlTwCF3DV2WSy5iCG9SKyR-GZFPA6m166ENJ7wtyLY auth_event_ids(7)=['$B2XdB4JXj97qtVbvHN0OJr7kKdtDVDwYGCPpHdU-GsI', '$sd7XJzwB6lFcy_pC512072QWlexAy1P4H0MDWFjQD5s', '$fV42Jv1SCvuMwv9Sj0pmdZhV9u0WaxEs1tMqs-FGSrY', '$-YRnU5-ieM5027RC21oG1US9QwAsmLMzq0SdJzvyDTg', '$Q55_Jem_AB_zkZIHVUpnPMU6Acl1n57YQMB4Y_o1rJQ', '$Z_kM2eubQ4zJpx4xT2hvo7s7-vDzl1blyrS8z05v5Js', '$yg2qSK4t4I7_PaPAjbf5kqmIGZobHnu4yjzK2BpBq2M']
2021-05-01 04:14:15,365 - synapse.rest.client.v1.room - 411 - INFO - POST-12 - /bulksend double_check_event_id=$uXlTwCF3DV2WSy5iCG9SKyR-GZFPA6m166ENJ7wtyLY auth_event_ids(7)=['$B2XdB4JXj97qtVbvHN0OJr7kKdtDVDwYGCPpHdU-GsI', '$sd7XJzwB6lFcy_pC512072QWlexAy1P4H0MDWFjQD5s', '$fV42Jv1SCvuMwv9Sj0pmdZhV9u0WaxEs1tMqs-FGSrY', '$-YRnU5-ieM5027RC21oG1US9QwAsmLMzq0SdJzvyDTg', '$Q55_Jem_AB_zkZIHVUpnPMU6Acl1n57YQMB4Y_o1rJQ', '$Z_kM2eubQ4zJpx4xT2hvo7s7-vDzl1blyrS8z05v5Js', '$yg2qSK4t4I7_PaPAjbf5kqmIGZobHnu4yjzK2BpBq2M']
2021-05-01 04:14:15,366 - synapse.logging.context - 70 - WARNING - POST-12 - Re-starting finished log context POST-12
2021-05-01 04:14:15,367 - synapse.logging.context - 70 - WARNING - POST-12 - Re-starting finished log context POST-12
2021-05-01 04:14:15,371 - synapse.logging.context - 70 - WARNING - POST-12 - Re-starting finished log context POST-12
2021-05-01 04:14:15,384 - synapse.rest.client.v1.room - 402 - INFO - POST-12 - /bulksend event_id=$e_fZOPvlWVTfCnckdZT8zuiT_zArA7rHgf5QKconAc8 auth_event_ids(7)=['$B2XdB4JXj97qtVbvHN0OJr7kKdtDVDwYGCPpHdU-GsI', '$sd7XJzwB6lFcy_pC512072QWlexAy1P4H0MDWFjQD5s', '$fV42Jv1SCvuMwv9Sj0pmdZhV9u0WaxEs1tMqs-FGSrY', '$-YRnU5-ieM5027RC21oG1US9QwAsmLMzq0SdJzvyDTg', '$Q55_Jem_AB_zkZIHVUpnPMU6Acl1n57YQMB4Y_o1rJQ', '$Z_kM2eubQ4zJpx4xT2hvo7s7-vDzl1blyrS8z05v5Js', '$yg2qSK4t4I7_PaPAjbf5kqmIGZobHnu4yjzK2BpBq2M']
2021-05-01 04:14:15,385 - synapse.rest.client.v1.room - 411 - INFO - POST-12 - /bulksend double_check_event_id=$e_fZOPvlWVTfCnckdZT8zuiT_zArA7rHgf5QKconAc8 auth_event_ids(7)=['$B2XdB4JXj97qtVbvHN0OJr7kKdtDVDwYGCPpHdU-GsI', '$sd7XJzwB6lFcy_pC512072QWlexAy1P4H0MDWFjQD5s', '$fV42Jv1SCvuMwv9Sj0pmdZhV9u0WaxEs1tMqs-FGSrY', '$-YRnU5-ieM5027RC21oG1US9QwAsmLMzq0SdJzvyDTg', '$Q55_Jem_AB_zkZIHVUpnPMU6Acl1n57YQMB4Y_o1rJQ', '$Z_kM2eubQ4zJpx4xT2hvo7s7-vDzl1blyrS8z05v5Js', '$yg2qSK4t4I7_PaPAjbf5kqmIGZobHnu4yjzK2BpBq2M']
```



## `hs2`

```
2021-05-01 04:14:16,637 - synapse.handlers.federation - 926 - INFO - GET-4 - backfill response returned 13 events
2021-05-01 04:14:16,638 - synapse.handlers.federation - 960 - INFO - GET-4 - backfill: Got 12 events with 0 edges
2021-05-01 04:14:16,646 - synapse.handlers.federation - 2431 - INFO - GET-4 - event.event_id=$B2XdB4JXj97qtVbvHN0OJr7kKdtDVDwYGCPpHdU-GsI event_auth_events=set()
2021-05-01 04:14:16,658 - synapse.handlers.federation - 2431 - INFO - GET-4 - event.event_id=$Q55_Jem_AB_zkZIHVUpnPMU6Acl1n57YQMB4Y_o1rJQ event_auth_events={'$B2XdB4JXj97qtVbvHN0OJr7kKdtDVDwYGCPpHdU-GsI'}
2021-05-01 04:14:16,669 - synapse.handlers.federation - 2431 - INFO - GET-4 - event.event_id=$sd7XJzwB6lFcy_pC512072QWlexAy1P4H0MDWFjQD5s event_auth_events={'$Q55_Jem_AB_zkZIHVUpnPMU6Acl1n57YQMB4Y_o1rJQ', '$B2XdB4JXj97qtVbvHN0OJr7kKdtDVDwYGCPpHdU-GsI'}
2021-05-01 04:14:16,680 - synapse.handlers.federation - 2431 - INFO - GET-4 - event.event_id=$fV42Jv1SCvuMwv9Sj0pmdZhV9u0WaxEs1tMqs-FGSrY event_auth_events={'$B2XdB4JXj97qtVbvHN0OJr7kKdtDVDwYGCPpHdU-GsI', '$Q55_Jem_AB_zkZIHVUpnPMU6Acl1n57YQMB4Y_o1rJQ', '$sd7XJzwB6lFcy_pC512072QWlexAy1P4H0MDWFjQD5s'}
2021-05-01 04:14:16,690 - synapse.handlers.federation - 2431 - INFO - GET-4 - event.event_id=$-YRnU5-ieM5027RC21oG1US9QwAsmLMzq0SdJzvyDTg event_auth_events={'$B2XdB4JXj97qtVbvHN0OJr7kKdtDVDwYGCPpHdU-GsI', '$Q55_Jem_AB_zkZIHVUpnPMU6Acl1n57YQMB4Y_o1rJQ', '$sd7XJzwB6lFcy_pC512072QWlexAy1P4H0MDWFjQD5s'}
2021-05-01 04:14:16,699 - synapse.handlers.federation - 2431 - INFO - GET-4 - event.event_id=$Z_kM2eubQ4zJpx4xT2hvo7s7-vDzl1blyrS8z05v5Js event_auth_events={'$fV42Jv1SCvuMwv9Sj0pmdZhV9u0WaxEs1tMqs-FGSrY', '$sd7XJzwB6lFcy_pC512072QWlexAy1P4H0MDWFjQD5s', '$B2XdB4JXj97qtVbvHN0OJr7kKdtDVDwYGCPpHdU-GsI'}
2021-05-01 04:14:16,705 - synapse.handlers.federation - 2431 - INFO - GET-4 - event.event_id=$QWV5wDAv_0JtTk81twPMMOCmjoN-T1egDbJs8WmuAOg event_auth_events={'$B2XdB4JXj97qtVbvHN0OJr7kKdtDVDwYGCPpHdU-GsI', '$Z_kM2eubQ4zJpx4xT2hvo7s7-vDzl1blyrS8z05v5Js', '$sd7XJzwB6lFcy_pC512072QWlexAy1P4H0MDWFjQD5s'}
2021-05-01 04:14:16,716 - synapse.handlers.federation - 2431 - INFO - GET-4 - event.event_id=$e_fZOPvlWVTfCnckdZT8zuiT_zArA7rHgf5QKconAc8 event_auth_events={'$sd7XJzwB6lFcy_pC512072QWlexAy1P4H0MDWFjQD5s', '$Q55_Jem_AB_zkZIHVUpnPMU6Acl1n57YQMB4Y_o1rJQ', '$-YRnU5-ieM5027RC21oG1US9QwAsmLMzq0SdJzvyDTg', '$Z_kM2eubQ4zJpx4xT2hvo7s7-vDzl1blyrS8z05v5Js', '$B2XdB4JXj97qtVbvHN0OJr7kKdtDVDwYGCPpHdU-GsI', '$yg2qSK4t4I7_PaPAjbf5kqmIGZobHnu4yjzK2BpBq2M', '$fV42Jv1SCvuMwv9Sj0pmdZhV9u0WaxEs1tMqs-FGSrY'}
2021-05-01 04:14:16,717 - synapse.handlers.federation - 2451 - INFO - GET-4 - auth_events contains unknown events: {'$yg2qSK4t4I7_PaPAjbf5kqmIGZobHnu4yjzK2BpBq2M'}

2021-05-01 04:14:16,727 - synapse.http.matrixfederationclient - 209 - INFO - GET-4 - {GET-O-7} [hs1] Completed request: 200 OK in 0.01 secs - GET matrix://hs1/_matrix/federation/v1/event_auth/%21nTiCFGsvijiehxacea%3Ahs1/%24e_fZOPvlWVTfCnckdZT8zuiT_zArA7rHgf5QKconAc8
2021-05-01 04:14:16,731 - synapse.federation.federation_base - 126 - WARNING - GET-4 - Signature check failed for $HJ1S9pSyk0IayKDIUx1o8YseQDhR5OqwDSLaWAhF0T8: 403: event id $HJ1S9pSyk0IayKDIUx1o8YseQDhR5OqwDSLaWAhF0T8: unable to verify signature for sender hs1: 401: Invalid signature for server hs1 with key ed25519:a_vRWK: Unable to verify signature for hs1: <class 'nacl.exceptions.BadSignatureError'> Signature was forged or corrupt
2021-05-01 04:14:16,733 - synapse.federation.federation_client - 417 - WARNING - GET-4 - Failed to find copy of $HJ1S9pSyk0IayKDIUx1o8YseQDhR5OqwDSLaWAhF0T8 with valid signature

2021-05-01 04:14:16,734 - synapse.federation.federation_client - 450 - INFO - GET-4 - get_event_auth auth_chain(7)=[<FrozenEventV3 event_id='$Z_kM2eubQ4zJpx4xT2hvo7s7-vDzl1blyrS8z05v5Js', type='m.room.member', state_key='@alice:hs1'>, <FrozenEventV3 event_id='$-YRnU5-ieM5027RC21oG1US9QwAsmLMzq0SdJzvyDTg', type='m.room.history_visibility', state_key=''>, <FrozenEventV3 event_id='$fV42Jv1SCvuMwv9Sj0pmdZhV9u0WaxEs1tMqs-FGSrY', type='m.room.join_rules', state_key=''>, <FrozenEventV3 event_id='$sd7XJzwB6lFcy_pC512072QWlexAy1P4H0MDWFjQD5s', type='m.room.power_levels', state_key=''>, <FrozenEventV3 event_id='$Q55_Jem_AB_zkZIHVUpnPMU6Acl1n57YQMB4Y_o1rJQ', type='m.room.member', state_key='@the-bridge-user:hs1'>, <FrozenEventV3 event_id='$HJ1S9pSyk0IayKDIUx1o8YseQDhR5OqwDSLaWAhF0T8', type='m.room.member', state_key='@maria:hs1'>, <FrozenEventV3 event_id='$B2XdB4JXj97qtVbvHN0OJr7kKdtDVDwYGCPpHdU-GsI', type='m.room.create', state_key=''>]

signed_auth(6)=[<FrozenEventV3 event_id='$Z_kM2eubQ4zJpx4xT2hvo7s7-vDzl1blyrS8z05v5Js', type='m.room.member', state_key='@alice:hs1'>, <FrozenEventV3 event_id='$-YRnU5-ieM5027RC21oG1US9QwAsmLMzq0SdJzvyDTg', type='m.room.history_visibility', state_key=''>, <FrozenEventV3 event_id='$fV42Jv1SCvuMwv9Sj0pmdZhV9u0WaxEs1tMqs-FGSrY', type='m.room.join_rules', state_key=''>, <FrozenEventV3 event_id='$sd7XJzwB6lFcy_pC512072QWlexAy1P4H0MDWFjQD5s', type='m.room.power_levels', state_key=''>, <FrozenEventV3 event_id='$Q55_Jem_AB_zkZIHVUpnPMU6Acl1n57YQMB4Y_o1rJQ', type='m.room.member', state_key='@the-bridge-user:hs1'>, <FrozenEventV3 event_id='$B2XdB4JXj97qtVbvHN0OJr7kKdtDVDwYGCPpHdU-GsI', type='m.room.create', state_key=''>]


2021-05-01 04:14:16,734 - synapse.handlers.federation - 2457 - INFO - GET-4 - auth_events retrieved remote_auth_chain=[<FrozenEventV3 event_id='$B2XdB4JXj97qtVbvHN0OJr7kKdtDVDwYGCPpHdU-GsI', type='m.room.create', state_key=''>, <FrozenEventV3 event_id='$Q55_Jem_AB_zkZIHVUpnPMU6Acl1n57YQMB4Y_o1rJQ', type='m.room.member', state_key='@the-bridge-user:hs1'>, <FrozenEventV3 event_id='$sd7XJzwB6lFcy_pC512072QWlexAy1P4H0MDWFjQD5s', type='m.room.power_levels', state_key=''>, <FrozenEventV3 event_id='$fV42Jv1SCvuMwv9Sj0pmdZhV9u0WaxEs1tMqs-FGSrY', type='m.room.join_rules', state_key=''>, <FrozenEventV3 event_id='$-YRnU5-ieM5027RC21oG1US9QwAsmLMzq0SdJzvyDTg', type='m.room.history_visibility', state_key=''>, <FrozenEventV3 event_id='$Z_kM2eubQ4zJpx4xT2hvo7s7-vDzl1blyrS8z05v5Js', type='m.room.member', state_key='@alice:hs1'>]
2021-05-01 04:14:16,735 - synapse.handlers.federation - 2469 - INFO - GET-4 - auth_events seen_remotes={'$Q55_Jem_AB_zkZIHVUpnPMU6Acl1n57YQMB4Y_o1rJQ', '$sd7XJzwB6lFcy_pC512072QWlexAy1P4H0MDWFjQD5s', '$-YRnU5-ieM5027RC21oG1US9QwAsmLMzq0SdJzvyDTg', '$Z_kM2eubQ4zJpx4xT2hvo7s7-vDzl1blyrS8z05v5Js', '$B2XdB4JXj97qtVbvHN0OJr7kKdtDVDwYGCPpHdU-GsI', '$fV42Jv1SCvuMwv9Sj0pmdZhV9u0WaxEs1tMqs-FGSrY'}
2021-05-01 04:14:16,735 - synapse.handlers.federation - 2516 - INFO - GET-4 - auth_events refers to events which are not in our calculated auth chain: {'$yg2qSK4t4I7_PaPAjbf5kqmIGZobHnu4yjzK2BpBq2M', '$sd7XJzwB6lFcy_pC512072QWlexAy1P4H0MDWFjQD5s', '$Q55_Jem_AB_zkZIHVUpnPMU6Acl1n57YQMB4Y_o1rJQ', '$-YRnU5-ieM5027RC21oG1US9QwAsmLMzq0SdJzvyDTg', '$fV42Jv1SCvuMwv9Sj0pmdZhV9u0WaxEs1tMqs-FGSrY', '$Z_kM2eubQ4zJpx4xT2hvo7s7-vDzl1blyrS8z05v5Js', '$B2XdB4JXj97qtVbvHN0OJr7kKdtDVDwYGCPpHdU-GsI'}
2021-05-01 04:14:16,736 - synapse.state - 452 - INFO - GET-4 - Resolving state for !nTiCFGsvijiehxacea:hs1 with 2 groups
2021-05-01 04:14:16,739 - synapse.handlers.federation - 2555 - INFO - GET-4 - After state res: updating auth_events with new state {('m.room.create', ''): '$B2XdB4JXj97qtVbvHN0OJr7kKdtDVDwYGCPpHdU-GsI', ('m.room.member', '@the-bridge-user:hs1'): '$Q55_Jem_AB_zkZIHVUpnPMU6Acl1n57YQMB4Y_o1rJQ', ('m.room.power_levels', ''): '$sd7XJzwB6lFcy_pC512072QWlexAy1P4H0MDWFjQD5s', ('m.room.join_rules', ''): '$fV42Jv1SCvuMwv9Sj0pmdZhV9u0WaxEs1tMqs-FGSrY', ('m.room.history_visibility', ''): '$-YRnU5-ieM5027RC21oG1US9QwAsmLMzq0SdJzvyDTg', ('m.room.member', '@alice:hs1'): '$Z_kM2eubQ4zJpx4xT2hvo7s7-vDzl1blyrS8z05v5Js'}
2021-05-01 04:14:16,742 - synapse.handlers.federation - 2389 - WARNING - GET-4 - Failed auth resolution for <FrozenEventV3 event_id='$e_fZOPvlWVTfCnckdZT8zuiT_zArA7rHgf5QKconAc8', type='m.room.message', state_key=None> because 403: User @maria:hs1 not in room !nTiCFGsvijiehxacea:hs1 (None)
2021-05-01 04:14:16,746 - synapse.handlers.federation - 2431 - INFO - GET-4 - event.event_id=$uXlTwCF3DV2WSy5iCG9SKyR-GZFPA6m166ENJ7wtyLY event_auth_events={'$sd7XJzwB6lFcy_pC512072QWlexAy1P4H0MDWFjQD5s', '$Q55_Jem_AB_zkZIHVUpnPMU6Acl1n57YQMB4Y_o1rJQ', '$-YRnU5-ieM5027RC21oG1US9QwAsmLMzq0SdJzvyDTg', '$Z_kM2eubQ4zJpx4xT2hvo7s7-vDzl1blyrS8z05v5Js', '$B2XdB4JXj97qtVbvHN0OJr7kKdtDVDwYGCPpHdU-GsI', '$yg2qSK4t4I7_PaPAjbf5kqmIGZobHnu4yjzK2BpBq2M', '$fV42Jv1SCvuMwv9Sj0pmdZhV9u0WaxEs1tMqs-FGSrY'}
2021-05-01 04:14:16,747 - synapse.handlers.federation - 2451 - INFO - GET-4 - auth_events contains unknown events: {'$yg2qSK4t4I7_PaPAjbf5kqmIGZobHnu4yjzK2BpBq2M'}
2021-05-01 04:14:16,754 - synapse.http.matrixfederationclient - 209 - INFO - GET-4 - {GET-O-8} [hs1] Completed request: 200 OK in 0.01 secs - GET matrix://hs1/_matrix/federation/v1/event_auth/%21nTiCFGsvijiehxacea%3Ahs1/%24uXlTwCF3DV2WSy5iCG9SKyR-GZFPA6m166ENJ7wtyLY
2021-05-01 04:14:16,756 - synapse.federation.federation_base - 126 - WARNING - GET-4 - Signature check failed for $HJ1S9pSyk0IayKDIUx1o8YseQDhR5OqwDSLaWAhF0T8: 403: event id $HJ1S9pSyk0IayKDIUx1o8YseQDhR5OqwDSLaWAhF0T8: unable to verify signature for sender hs1: 401: Invalid signature for server hs1 with key ed25519:a_vRWK: Unable to verify signature for hs1: <class 'nacl.exceptions.BadSignatureError'> Signature was forged or corrupt
2021-05-01 04:14:16,759 - synapse.federation.federation_client - 417 - WARNING - GET-4 - Failed to find copy of $HJ1S9pSyk0IayKDIUx1o8YseQDhR5OqwDSLaWAhF0T8 with valid signature
2021-05-01 04:14:16,759 - synapse.federation.federation_client - 450 - INFO - GET-4 - get_event_auth auth_chain(7)=[<FrozenEventV3 event_id='$Z_kM2eubQ4zJpx4xT2hvo7s7-vDzl1blyrS8z05v5Js', type='m.room.member', state_key='@alice:hs1'>, <FrozenEventV3 event_id='$-YRnU5-ieM5027RC21oG1US9QwAsmLMzq0SdJzvyDTg', type='m.room.history_visibility', state_key=''>, <FrozenEventV3 event_id='$fV42Jv1SCvuMwv9Sj0pmdZhV9u0WaxEs1tMqs-FGSrY', type='m.room.join_rules', state_key=''>, <FrozenEventV3 event_id='$sd7XJzwB6lFcy_pC512072QWlexAy1P4H0MDWFjQD5s', type='m.room.power_levels', state_key=''>, <FrozenEventV3 event_id='$Q55_Jem_AB_zkZIHVUpnPMU6Acl1n57YQMB4Y_o1rJQ', type='m.room.member', state_key='@the-bridge-user:hs1'>, <FrozenEventV3 event_id='$HJ1S9pSyk0IayKDIUx1o8YseQDhR5OqwDSLaWAhF0T8', type='m.room.member', state_key='@maria:hs1'>, <FrozenEventV3 event_id='$B2XdB4JXj97qtVbvHN0OJr7kKdtDVDwYGCPpHdU-GsI', type='m.room.create', state_key=''>] signed_auth(6)=[<FrozenEventV3 event_id='$Z_kM2eubQ4zJpx4xT2hvo7s7-vDzl1blyrS8z05v5Js', type='m.room.member', state_key='@alice:hs1'>, <FrozenEventV3 event_id='$-YRnU5-ieM5027RC21oG1US9QwAsmLMzq0SdJzvyDTg', type='m.room.history_visibility', state_key=''>, <FrozenEventV3 event_id='$fV42Jv1SCvuMwv9Sj0pmdZhV9u0WaxEs1tMqs-FGSrY', type='m.room.join_rules', state_key=''>, <FrozenEventV3 event_id='$sd7XJzwB6lFcy_pC512072QWlexAy1P4H0MDWFjQD5s', type='m.room.power_levels', state_key=''>, <FrozenEventV3 event_id='$Q55_Jem_AB_zkZIHVUpnPMU6Acl1n57YQMB4Y_o1rJQ', type='m.room.member', state_key='@the-bridge-user:hs1'>, <FrozenEventV3 event_id='$B2XdB4JXj97qtVbvHN0OJr7kKdtDVDwYGCPpHdU-GsI', type='m.room.create', state_key=''>]
2021-05-01 04:14:16,759 - synapse.handlers.federation - 2457 - INFO - GET-4 - auth_events retrieved remote_auth_chain=[<FrozenEventV3 event_id='$B2XdB4JXj97qtVbvHN0OJr7kKdtDVDwYGCPpHdU-GsI', type='m.room.create', state_key=''>, <FrozenEventV3 event_id='$Q55_Jem_AB_zkZIHVUpnPMU6Acl1n57YQMB4Y_o1rJQ', type='m.room.member', state_key='@the-bridge-user:hs1'>, <FrozenEventV3 event_id='$sd7XJzwB6lFcy_pC512072QWlexAy1P4H0MDWFjQD5s', type='m.room.power_levels', state_key=''>, <FrozenEventV3 event_id='$fV42Jv1SCvuMwv9Sj0pmdZhV9u0WaxEs1tMqs-FGSrY', type='m.room.join_rules', state_key=''>, <FrozenEventV3 event_id='$-YRnU5-ieM5027RC21oG1US9QwAsmLMzq0SdJzvyDTg', type='m.room.history_visibility', state_key=''>, <FrozenEventV3 event_id='$Z_kM2eubQ4zJpx4xT2hvo7s7-vDzl1blyrS8z05v5Js', type='m.room.member', state_key='@alice:hs1'>]
2021-05-01 04:14:16,760 - synapse.metrics - 587 - INFO - sentinel - Collecting gc 1
2021-05-01 04:14:16,761 - synapse.handlers.federation - 2469 - INFO - GET-4 - auth_events seen_remotes={'$Q55_Jem_AB_zkZIHVUpnPMU6Acl1n57YQMB4Y_o1rJQ', '$sd7XJzwB6lFcy_pC512072QWlexAy1P4H0MDWFjQD5s', '$-YRnU5-ieM5027RC21oG1US9QwAsmLMzq0SdJzvyDTg', '$Z_kM2eubQ4zJpx4xT2hvo7s7-vDzl1blyrS8z05v5Js', '$B2XdB4JXj97qtVbvHN0OJr7kKdtDVDwYGCPpHdU-GsI', '$fV42Jv1SCvuMwv9Sj0pmdZhV9u0WaxEs1tMqs-FGSrY'}
2021-05-01 04:14:16,761 - synapse.handlers.federation - 2516 - INFO - GET-4 - auth_events refers to events which are not in our calculated auth chain: {'$yg2qSK4t4I7_PaPAjbf5kqmIGZobHnu4yjzK2BpBq2M', '$Q55_Jem_AB_zkZIHVUpnPMU6Acl1n57YQMB4Y_o1rJQ', '$-YRnU5-ieM5027RC21oG1US9QwAsmLMzq0SdJzvyDTg', '$fV42Jv1SCvuMwv9Sj0pmdZhV9u0WaxEs1tMqs-FGSrY', '$Z_kM2eubQ4zJpx4xT2hvo7s7-vDzl1blyrS8z05v5Js'}
2021-05-01 04:14:16,762 - synapse.state - 452 - INFO - GET-4 - Resolving state for !nTiCFGsvijiehxacea:hs1 with 2 groups
2021-05-01 04:14:16,764 - synapse.handlers.federation - 2555 - INFO - GET-4 - After state res: updating auth_events with new state {('m.room.member', '@the-bridge-user:hs1'): '$Q55_Jem_AB_zkZIHVUpnPMU6Acl1n57YQMB4Y_o1rJQ', ('m.room.join_rules', ''): '$fV42Jv1SCvuMwv9Sj0pmdZhV9u0WaxEs1tMqs-FGSrY', ('m.room.history_visibility', ''): '$-YRnU5-ieM5027RC21oG1US9QwAsmLMzq0SdJzvyDTg', ('m.room.member', '@alice:hs1'): '$Z_kM2eubQ4zJpx4xT2hvo7s7-vDzl1blyrS8z05v5Js'}
2021-05-01 04:14:16,768 - synapse.handlers.federation - 2389 - WARNING - GET-4 - Failed auth resolution for <FrozenEventV3 event_id='$uXlTwCF3DV2WSy5iCG9SKyR-GZFPA6m166ENJ7wtyLY', type='m.room.message', state_key=None> because 403: User @maria:hs1 not in room !nTiCFGsvijiehxacea:hs1 (None)
2021-05-01 04:14:16,773 - synapse.handlers.federation - 2431 - INFO - GET-4 - event.event_id=$_8vQdgFMafAaWskZoYTdvKFMp44hvS50Myp1DoMyMRg event_auth_events={'$B2XdB4JXj97qtVbvHN0OJr7kKdtDVDwYGCPpHdU-GsI', '$Z_kM2eubQ4zJpx4xT2hvo7s7-vDzl1blyrS8z05v5Js', '$sd7XJzwB6lFcy_pC512072QWlexAy1P4H0MDWFjQD5s'}
2021-05-01 04:14:16,778 - synapse.handlers.federation - 2431 - INFO - GET-4 - event.event_id=$MzJyjtpWn1VUYgwU_KDKxaXP-4A-SFzl5aPdCpoMtIY event_auth_events={'$B2XdB4JXj97qtVbvHN0OJr7kKdtDVDwYGCPpHdU-GsI', '$Z_kM2eubQ4zJpx4xT2hvo7s7-vDzl1blyrS8z05v5Js', '$sd7XJzwB6lFcy_pC512072QWlexAy1P4H0MDWFjQD5s'}
2021-05-01 04:14:16,785 - synapse.handlers.federation - 2431 - INFO - GET-4 - event.event_id=$3zmsrY2NV1lZuTST2DjsAYnf_n61mj-F7FzwvOaurbM event_auth_events={'$B2XdB4JXj97qtVbvHN0OJr7kKdtDVDwYGCPpHdU-GsI', '$Z_kM2eubQ4zJpx4xT2hvo7s7-vDzl1blyrS8z05v5Js', '$sd7XJzwB6lFcy_pC512072QWlexAy1P4H0MDWFjQD5s'}
```


</details>

Reply:
It looks like the random `event_id` is coming from when we try to parse the JSON events from `/event_auth` which don't include an `event_id`. So we end up with a generated one from the `compute_event_reference_hash(ev)`, [`synapse/events/__init__.py#L482-L484`](https://github.com/matrix-org/synapse/blob/4d624f467a6c252b9295e46b8b83f36b1f0d3d45/synapse/events/__init__.py#L482-L484). The hash ends up being the same for the other events but it's wrong with the `@maria:hs1` event 🤔

Reply:
From https://github.com/matrix-org/synapse/issues/3294, I was wondering if `Signature was forged or corrupt` was a [URL canonicalization problem and `%xx` escapes with the reverse proxy](https://github.com/matrix-org/synapse/blob/develop/docs/reverse_proxy.md) but the event in URL is getting parsed fine and I only see a [reverse proxy used in Complement with the SynapseWorkers dockerfile](https://github.com/matrix-org/complement/blob/8dc7464235912b1ed5324ee722ef5762437aae07/dockerfiles/synapse/caddy.complement.json#L25). Only having a signature problem with one event from a response.

Reply:
Related issue, https://github.com/matrix-org/synapse/issues/4012, not sure if relevant.

Reply:
@erikjohnston What are the key things to look out for with the signatures? I'm going to dive more into this tomorrow.

Reply:
It looks like the signature check discrepancy is coming from having different `auth_events` for the `@maria:hs1` join event when it is signed vs when it is verified. 

```
2021-05-05 04:16:21,823 - signedjson.sign - 62 - INFO - POST-12 - SIGNING:
signature_name=hs1
key_id=ed25519:a_jTiH
message_bytes=b'{"auth_events":["$WbbbFKpc5jykjXVJ6m9eO8gfYMmwT50PzUQktCjw13Y","$Ho6pa-Sojs4ITAvbpl4ZD2E5z8AepnveZa7m3_9qumA","$XA9a9WFaeXN6PHnIx_uFuGQaRgCv-hwX-47RC4jezIE","$mPy6iAqn0liZtRA0Mce_3x5DJy5rv35Ip1V-wMOCly0","$IS4vnImsZswBX0_OYoIJcFU9XQfxGOdbCu3xreniqFo","$UKsTBHyn0Lq3OX-cv4pyjifEFMBfTYMqjvAlQwQAMXU"],"content":{"membership":"join"},"depth":1,"hashes":{"sha256":"K9uleje8Ad71GxHSo7Jya1W508qSofcQ4+tao56Jebw"},"origin":"hs1","origin_server_ts":1620188181822,"prev_events":["$tjfzZilJynbhBiZPSyaQndRdwYHwrLoFiylfIjWCVDo"],"prev_state":[],"room_id":"!tNBiindlQJqJOiGWFF:hs1","sender":"@maria:hs1","state_key":"@maria:hs1","type":"m.room.member"}'
encode_signing_key_base64=BWyBHEEX2l5RXodNpORTqeLFRoHOOKVAMTuyddNZ364

Pretty ->
{
  "auth_events": [
    "$WbbbFKpc5jykjXVJ6m9eO8gfYMmwT50PzUQktCjw13Y",
    "$Ho6pa-Sojs4ITAvbpl4ZD2E5z8AepnveZa7m3_9qumA",
    "$XA9a9WFaeXN6PHnIx_uFuGQaRgCv-hwX-47RC4jezIE",
    "$mPy6iAqn0liZtRA0Mce_3x5DJy5rv35Ip1V-wMOCly0",
    "$IS4vnImsZswBX0_OYoIJcFU9XQfxGOdbCu3xreniqFo",
    "$UKsTBHyn0Lq3OX-cv4pyjifEFMBfTYMqjvAlQwQAMXU"
  ],
  "content": {
    "membership": "join"
  },
  "depth": 1,
  "hashes": {
    "sha256": "K9uleje8Ad71GxHSo7Jya1W508qSofcQ4+tao56Jebw"
  },
  "origin": "hs1",
  "origin_server_ts": 1620188181822,
  "prev_events": [
    "$tjfzZilJynbhBiZPSyaQndRdwYHwrLoFiylfIjWCVDo"
  ],
  "prev_state": [],
  "room_id": "!tNBiindlQJqJOiGWFF:hs1",
  "sender": "@maria:hs1",
  "state_key": "@maria:hs1",
  "type": "m.room.member"
}
```



There is an extra `$2dsJ7pVhnrQ9E_lZ7YPapt7CIKENbNRhsOXxu9We6lk`  in the `auth_events` which appears to be referencing itself.
```diff
2021-05-05 04:16:23,132 - signedjson.sign - 143 - INFO - GET-4 - VERIFY:
signature_name=hs1
key_id=ed25519:a_jTiH
message=b'{"auth_events":["$WbbbFKpc5jykjXVJ6m9eO8gfYMmwT50PzUQktCjw13Y","$Ho6pa-Sojs4ITAvbpl4ZD2E5z8AepnveZa7m3_9qumA","$XA9a9WFaeXN6PHnIx_uFuGQaRgCv-hwX-47RC4jezIE","$mPy6iAqn0liZtRA0Mce_3x5DJy5rv35Ip1V-wMOCly0","$IS4vnImsZswBX0_OYoIJcFU9XQfxGOdbCu3xreniqFo","$UKsTBHyn0Lq3OX-cv4pyjifEFMBfTYMqjvAlQwQAMXU","$2dsJ7pVhnrQ9E_lZ7YPapt7CIKENbNRhsOXxu9We6lk"],"content":{"membership":"join"},"depth":1,"hashes":{"sha256":"K9uleje8Ad71GxHSo7Jya1W508qSofcQ4+tao56Jebw"},"origin":"hs1","origin_server_ts":1620188181822,"prev_events":["$tjfzZilJynbhBiZPSyaQndRdwYHwrLoFiylfIjWCVDo"],"prev_state":[],"room_id":"!tNBiindlQJqJOiGWFF:hs1","sender":"@maria:hs1","state_key":"@maria:hs1","type":"m.room.member"}' 
signature=b'E\xd9/\xfe\x1f\x80V|\xdc\x9d\x81\xd5\xa8c;<DT\x8f\xef \x9d\x82"\x17sgY\x18j\x91\xe9\x9c\x03\xbbY\x89\xf5\xfb\x16\xe1{\'P\n\xd8\x18\xa3&\x90\x83\xfc\xb0\xf1\xfb}\xe6\x89K5\xe6\x98\x00\x06' 
encode_verify_key_base64=stmQUQhX4yCDSw8hUm96cVFhaqM4Xh+ZCb044MF2tpk

Pretty ->
{
  "auth_events": [
    "$WbbbFKpc5jykjXVJ6m9eO8gfYMmwT50PzUQktCjw13Y",
    "$Ho6pa-Sojs4ITAvbpl4ZD2E5z8AepnveZa7m3_9qumA",
    "$XA9a9WFaeXN6PHnIx_uFuGQaRgCv-hwX-47RC4jezIE",
    "$mPy6iAqn0liZtRA0Mce_3x5DJy5rv35Ip1V-wMOCly0",
    "$IS4vnImsZswBX0_OYoIJcFU9XQfxGOdbCu3xreniqFo",
    "$UKsTBHyn0Lq3OX-cv4pyjifEFMBfTYMqjvAlQwQAMXU",
+    "$2dsJ7pVhnrQ9E_lZ7YPapt7CIKENbNRhsOXxu9We6lk"
  ],
  "content": {
    "membership": "join"
  },
  "depth": 1,
  "hashes": {
    "sha256": "K9uleje8Ad71GxHSo7Jya1W508qSofcQ4+tao56Jebw"
  },
  "origin": "hs1",
  "origin_server_ts": 1620188181822,
  "prev_events": [
    "$tjfzZilJynbhBiZPSyaQndRdwYHwrLoFiylfIjWCVDo"
  ],
  "prev_state": [],
  "room_id": "!tNBiindlQJqJOiGWFF:hs1",
  "sender": "@maria:hs1",
  "state_key": "@maria:hs1",
  "type": "m.room.member"
}
```

Reply:
That's very strange. My best guess is that the event is being modified after creation, and so the generated event ID is wrong? Are you copying an existing membership event and mutating it? Though I'd expect things to explode if you tried to do that.

(FTR this is technically about computing a hash rather than a signature for the event, but that's a minor difference)

Reply:
> (FTR this is technically about computing a hash rather than a signature for the event, but that's a minor difference)

🤷‍♀️ whatever is stored in `signatures` and using `signedjson` to sign/verify is mismatching.


>  Are you copying an existing membership event and mutating it? Though I'd expect things to explode if you tried to do that.

No, it's created from `room_member_handler.update_membership(...)`

---



After the test is ran, if I check the database, the `event_json` table has the proper expected 6 `auth_events` defined. If I check the `event_auth` table, it also lists 6 as expected.

```
sqlite> select * from event_json where event_id='$LfK8zWi0g_snvqwi93vWGKj-iTf7gxWh_WRhTD5pALc';
$LfK8zWi0g_snvqwi93vWGKj-iTf7gxWh_WRhTD5pALc|!AHbrvRDqXrkCVfXKhl:hs1|{"outlier":true}|{"auth_events":["$xHTb-UXboT1JvH3-RTRo0GvQsiIUCCbdXKgCFENvJ0M","$hXSwY12xa1QFdLhtakmKUTHcGqBZxYfMGScoy9YK7qA","$dR0n0X3XO39At9Dnh0n9R7yP2cACixg_M62unJwfghk","$Oq65vbSnXivje5Vn7StwlbC05JNxbspYUt2QEg0mevs","$iQ57FdhPGauzJa5EcKQMMyXDM-HEp6qpMUtq3lEeOv0","$qYQz3YSxMtLY4YmpCLH4jELozUON4_kp4BbEKR30dZ0"],"prev_events":["$NGICpLXTJXAApbDjbOTFNbUPPtZZeWWYmTYHKKQqcYu"],"type":"m.room.member","room_id":"!AHbrvRDqXrkCVfXKhl:hs1","sender":"@maria:hs1","content":{"membership":"join","displayname":"maria"},"depth":1,"prev_state":[],"state_key":"@maria:hs1","origin":"hs1","origin_server_ts":1620263077435,"hashes":{"sha256":"SG6DoLw/GBnzlHmAAkWrfYI549wfsq4rwtw8hfDUH+M"},"signatures":{"hs1":{"ed25519:a_epYV":"nKFH1vgrzi4uIhQ1V12HwYo1B2HpLSutjhZoTuCsm94J46ceRVgWU+9AaYHp55IJi27e7hCDmSU9TNSDM+wIBg"}},"unsigned":{"age_ts":1620263077435}}|3
```

```
sqlite> select * from event_auth where event_id='$LfK8zWi0g_snvqwi93vWGKj-iTf7gxWh_WRhTD5pALc';
$LfK8zWi0g_snvqwi93vWGKj-iTf7gxWh_WRhTD5pALc|$xHTb-UXboT1JvH3-RTRo0GvQsiIUCCbdXKgCFENvJ0M|!AHbrvRDqXrkCVfXKhl:hs1
$LfK8zWi0g_snvqwi93vWGKj-iTf7gxWh_WRhTD5pALc|$hXSwY12xa1QFdLhtakmKUTHcGqBZxYfMGScoy9YK7qA|!AHbrvRDqXrkCVfXKhl:hs1
$LfK8zWi0g_snvqwi93vWGKj-iTf7gxWh_WRhTD5pALc|$dR0n0X3XO39At9Dnh0n9R7yP2cACixg_M62unJwfghk|!AHbrvRDqXrkCVfXKhl:hs1
$LfK8zWi0g_snvqwi93vWGKj-iTf7gxWh_WRhTD5pALc|$Oq65vbSnXivje5Vn7StwlbC05JNxbspYUt2QEg0mevs|!AHbrvRDqXrkCVfXKhl:hs1
$LfK8zWi0g_snvqwi93vWGKj-iTf7gxWh_WRhTD5pALc|$iQ57FdhPGauzJa5EcKQMMyXDM-HEp6qpMUtq3lEeOv0|!AHbrvRDqXrkCVfXKhl:hs1
$LfK8zWi0g_snvqwi93vWGKj-iTf7gxWh_WRhTD5pALc|$qYQz3YSxMtLY4YmpCLH4jELozUON4_kp4BbEKR30dZ0|!AHbrvRDqXrkCVfXKhl:hs1
```

---

If I check the `_get_event_cache` wherever it is `set`, it sets it once properly as 6 `auth_events`

---


From the `/bulksend` endpoint where we create the historical join for `@maria:hs1`, the `auth_events_ids` are set correctly (6 of them). And if I do a separate lookup after that event is persisted, I see the proper amount of `auth_events_ids` (still 6).

But when the same join event is fetched in the `auth_chain` for another message, it has 7 `auth_event_ids`, the extra `auth_event` being itself.

Reply:
Ahhhh, it's because I `auth_event_ids.append(event_id)` and the reference is the same 🤦 so the entry in the cache is also updated because it references the same list.

Reply:
Woot! It's working over federation now 😌🎉🎉

```
--- PASS: TestBackfillingHistory (10.55s)
    --- PASS: TestBackfillingHistory/parallel (0.00s)
        --- PASS: TestBackfillingHistory/parallel/Historical_messages_are_visible_on_federated_server (1.82s)
```